File: sdks/typescript/src/core/security-tier.ts
/**
 * DCP v2.0 Adaptive Security Tier Engine.
 *
 * Automatically selects the appropriate cryptographic security tier based on
 * intent risk score, data classification, and action type. The tier drives
 * verification mode selection and PQ checkpoint intervals.
 *
 * Tier 0 (ROUTINE)  — Ed25519 only, PQ checkpoint every 50 events
 * Tier 1 (STANDARD)  — Ed25519 + PQ checkpoint every 10 events
 * Tier 2 (ELEVATED)  — Full hybrid per-operation, checkpoint every event
 * Tier 3 (MAXIMUM)   — Full hybrid + immediate checkpoint verification
 */

import type { IntentV2, VerificationMode, DataClass, SecurityTier } from '../types/v2.js';

export type { SecurityTier } from '../types/v2.js';

const SENSITIVE_DATA_CLASSES: ReadonlySet<DataClass> = new Set([
  'pii',
  'financial_data',
  'health_data',
  'credentials',
  'children_data',
]);

const HIGH_VALUE_DATA_CLASSES: ReadonlySet<DataClass> = new Set([
  'credentials',
  'children_data',
]);

const TIER_TO_VERIFICATION_MODE: Record<SecurityTier, VerificationMode> = {
  routine: 'classical_only',
  standard: 'hybrid_preferred',
  elevated: 'hybrid_required',
  maximum: 'hybrid_required',
};

const TIER_TO_CHECKPOINT_INTERVAL: Record<SecurityTier, number> = {
  routine: 50,
  standard: 10,
  elevated: 1,
  maximum: 1,
};

/**
 * Numeric ordering for tier comparison — higher value = stricter tier.
 */
const TIER_RANK: Record<SecurityTier, number> = {
  routine: 0,
  standard: 1,
  elevated: 2,
  maximum: 3,
};

/**
 * Compute the security tier for an intent based on risk score,
 * data classes, and action type. When risk_score is not provided,
 * classification falls back to data class / action type heuristics.
 */
export function computeSecurityTier(intent: IntentV2): SecurityTier {
  const score = (intent as IntentV2 & { risk_score?: number }).risk_score ?? 0;

  const hasHighValueData = intent.data_classes?.some((d) => HIGH_VALUE_DATA_CLASSES.has(d)) ?? false;
  const hasSensitiveData = intent.data_classes?.some((d) => SENSITIVE_DATA_CLASSES.has(d)) ?? false;
  const isPayment: boolean = intent.action_type === 'initiate_payment';

  if (score >= 800 || hasHighValueData) return 'maximum';
  if (score >= 500 || hasSensitiveData || isPayment) return 'elevated';
  if (score >= 200) return 'standard';
  return 'routine';
}

/**
 * Return the strictest (highest-rank) tier — useful when combining
 * auto-computed tier with an explicit override that must never downgrade.
 */
export function maxTier(a: SecurityTier, b: SecurityTier): SecurityTier {
  return TIER_RANK[a] >= TIER_RANK[b] ? a : b;
}

/**
 * Map a security tier to the corresponding verification mode.
 */
export function tierToVerificationMode(tier: SecurityTier): VerificationMode {
  return TIER_TO_VERIFICATION_MODE[tier];
}

/**
 * Map a security tier to the PQ checkpoint interval (number of events
 * between automatic checkpoints).
 */
export function tierToCheckpointInterval(tier: SecurityTier): number {
  return TIER_TO_CHECKPOINT_INTERVAL[tier];
}

File: sdks/python/dcp_ai/v2/security_tier.py
"""
DCP v2.0 Adaptive Security Tier Engine — Python port.
"""

from __future__ import annotations

from typing import Any

_SENSITIVE_DATA_CLASSES = frozenset(
    {"pii", "financial_data", "health_data", "credentials", "children_data"}
)
_HIGH_VALUE_DATA_CLASSES = frozenset({"credentials", "children_data"})

_TIER_TO_VERIFICATION_MODE = {
    "routine": "classical_only",
    "standard": "hybrid_preferred",
    "elevated": "hybrid_required",
    "maximum": "hybrid_required",
}
_TIER_TO_CHECKPOINT_INTERVAL = {
    "routine": 50,
    "standard": 10,
    "elevated": 1,
    "maximum": 1,
}
_TIER_RANK = {"routine": 0, "standard": 1, "elevated": 2, "maximum": 3}


def compute_security_tier(intent: dict[str, Any]) -> str:
    """Compute the security tier from an intent's risk score, data classes, and action type."""
    score = int(intent.get("risk_score", 0) or 0)
    data_classes = intent.get("data_classes") or []
    has_high_value = any(d in _HIGH_VALUE_DATA_CLASSES for d in data_classes)
    has_sensitive = any(d in _SENSITIVE_DATA_CLASSES for d in data_classes)
    is_payment = intent.get("action_type") == "initiate_payment"

    if score >= 800 or has_high_value:
        return "maximum"
    if score >= 500 or has_sensitive or is_payment:
        return "elevated"
    if score >= 200:
        return "standard"
    return "routine"


def max_tier(a: str, b: str) -> str:
    """Return the stricter of two tiers (used for floors that must never downgrade)."""
    return a if _TIER_RANK.get(a, 0) >= _TIER_RANK.get(b, 0) else b


def tier_to_verification_mode(tier: str) -> str:
    """Map a tier to the corresponding verification mode."""
    return _TIER_TO_VERIFICATION_MODE[tier]


def tier_to_checkpoint_interval(tier: str) -> int:
    """Map a tier to its PQ-checkpoint interval (number of events between checkpoints)."""
    return _TIER_TO_CHECKPOINT_INTERVAL[tier]
